File: applications/rtkforwardprojections/rtkforwardprojections.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkforwardprojections_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkThreeDCircularProjectionGeometryXMLFile.h"
#include "rtkJosephForwardProjectionImageFilter.h"
#include "rtkMaximumIntensityProjectionImageFilter.h"
#include "rtkJosephForwardAttenuatedProjectionImageFilter.h"
#include "rtkZengForwardProjectionImageFilter.h"
#ifdef RTK_USE_CUDA
#  include "rtkCudaForwardProjectionImageFilter.h"
#endif

#include <itkImageFileReader.h>
#include <itkImageFileWriter.h>

int
main(int argc, char * argv[])
{
  GGO(rtkforwardprojections, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

#ifdef RTK_USE_CUDA
  using OutputImageType = itk::CudaImage<OutputPixelType, Dimension>;
#else
  using OutputImageType = itk::Image<OutputPixelType, Dimension>;
#endif

  // Geometry
  if (args_info.verbose_flag)
    std::cout << "Reading geometry information from " << args_info.geometry_arg << "..." << std::flush;
  rtk::ThreeDCircularProjectionGeometry::Pointer geometry;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(geometry = rtk::ReadGeometry(args_info.geometry_arg));
  if (args_info.verbose_flag)
    std::cout << " done." << std::endl;

  // Create a stack of empty projection images
  using ConstantImageSourceType = rtk::ConstantImageSource<OutputImageType>;
  auto constantImageSource = ConstantImageSourceType::New();
  rtk::SetConstantImageSourceFromGgo<ConstantImageSourceType, args_info_rtkforwardprojections>(constantImageSource,
                                                                                               args_info);

  // Adjust size according to geometry
  auto sizeOutput = itk::MakeSize(
    constantImageSource->GetSize()[0], constantImageSource->GetSize()[1], geometry->GetGantryAngles().size());
  constantImageSource->SetSize(sizeOutput);

  // Input reader
  if (args_info.verbose_flag)
    std::cout << "Reading input volume " << args_info.input_arg << "..." << std::endl;
  OutputImageType::Pointer inputVolume;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(inputVolume = itk::ReadImage<OutputImageType>(args_info.input_arg))

  OutputImageType::Pointer attenuationMap;
  if (args_info.attenuationmap_given)
  {
    if (args_info.verbose_flag)
      std::cout << "Reading attenuation map " << args_info.attenuationmap_arg << "..." << std::endl;
    // Read an existing image to initialize the attenuation map
    attenuationMap = itk::ReadImage<OutputImageType>(args_info.attenuationmap_arg);
  }

  using ClipImageType = itk::Image<double, Dimension>;
  ClipImageType::Pointer inferiorClipImage, superiorClipImage;
  if (args_info.inferiorclipimage_given)
  {
    if (args_info.verbose_flag)
      std::cout << "Reading inferior clip image " << args_info.inferiorclipimage_arg << "..." << std::endl;
    // Read an existing image to initialize the attenuation map
    inferiorClipImage = itk::ReadImage<ClipImageType>(args_info.inferiorclipimage_arg);
  }
  if (args_info.superiorclipimage_given)
  {
    if (args_info.verbose_flag)
      std::cout << "Reading superior clip image " << args_info.superiorclipimage_arg << "..." << std::endl;
    // Read an existing image to initialize the attenuation map
    superiorClipImage = itk::ReadImage<ClipImageType>(args_info.superiorclipimage_arg);
  }

  // Create forward projection image filter
  if (args_info.verbose_flag)
    std::cout << "Projecting volume..." << std::endl;

  rtk::ForwardProjectionImageFilter<OutputImageType, OutputImageType>::Pointer forwardProjection;

  switch (args_info.fp_arg)
  {
    case (fp_arg_Joseph):
      forwardProjection = rtk::JosephForwardProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (fp_arg_JosephAttenuated):
      forwardProjection = rtk::JosephForwardAttenuatedProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (fp_arg_Zeng):
      forwardProjection = rtk::ZengForwardProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (fp_arg_MIP):
      forwardProjection = rtk::MaximumIntensityProjectionImageFilter<OutputImageType, OutputImageType>::New();
      break;
    case (fp_arg_CudaRayCast):
    {
#ifdef RTK_USE_CUDA
      forwardProjection = rtk::CudaForwardProjectionImageFilter<OutputImageType, OutputImageType>::New();
      auto * cudaForward = dynamic_cast<rtk::CudaForwardProjectionImageFilter<OutputImageType, OutputImageType> *>(
        forwardProjection.GetPointer());
      if (cudaForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to CudaForwardProjectionImageFilter." << std::endl;
        return EXIT_FAILURE;
      }
      cudaForward->SetStepSize(args_info.step_arg);
#else
      std::cerr << "The program has not been compiled with cuda option" << std::endl;
      return EXIT_FAILURE;
#endif
      break;
    }
    default:
      std::cerr << "Unhandled --method value." << std::endl;
      return EXIT_FAILURE;
  }
  forwardProjection->SetInput(constantImageSource->GetOutput());
  forwardProjection->SetInput(1, inputVolume);
  if (args_info.attenuationmap_given)
    forwardProjection->SetInput(2, attenuationMap);
  if (args_info.inferiorclipimage_given)
  {
    if (args_info.fp_arg == fp_arg_Joseph)
    {
      auto * josephForward = dynamic_cast<rtk::JosephForwardProjectionImageFilter<OutputImageType, OutputImageType> *>(
        forwardProjection.GetPointer());
      if (josephForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to JosephForwardProjectionImageFilter." << std::endl;
        return EXIT_FAILURE;
      }
      josephForward->SetInferiorClipImage(inferiorClipImage);
    }
    else if (args_info.fp_arg == fp_arg_JosephAttenuated)
    {
      auto * josephAttenuatedForward =
        dynamic_cast<rtk::JosephForwardAttenuatedProjectionImageFilter<OutputImageType, OutputImageType> *>(
          forwardProjection.GetPointer());
      if (josephAttenuatedForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to JosephForwardAttenuatedProjectionImageFilter."
                  << std::endl;
        return EXIT_FAILURE;
      }
      josephAttenuatedForward->SetInferiorClipImage(inferiorClipImage);
    }
    else if (args_info.fp_arg == fp_arg_MIP)
    {
      auto * mipForward = dynamic_cast<rtk::MaximumIntensityProjectionImageFilter<OutputImageType, OutputImageType> *>(
        forwardProjection.GetPointer());
      if (mipForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to MaximumIntensityProjectionImageFilter." << std::endl;
        return EXIT_FAILURE;
      }
      mipForward->SetInferiorClipImage(inferiorClipImage);
    }
  }
  if (args_info.superiorclipimage_given)
  {
    if (args_info.fp_arg == fp_arg_Joseph)
    {
      auto * josephForward = dynamic_cast<rtk::JosephForwardProjectionImageFilter<OutputImageType, OutputImageType> *>(
        forwardProjection.GetPointer());
      if (josephForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to JosephForwardProjectionImageFilter." << std::endl;
        return EXIT_FAILURE;
      }
      josephForward->SetSuperiorClipImage(superiorClipImage);
    }
    else if (args_info.fp_arg == fp_arg_JosephAttenuated)
    {
      auto * josephAttenuatedForward =
        dynamic_cast<rtk::JosephForwardAttenuatedProjectionImageFilter<OutputImageType, OutputImageType> *>(
          forwardProjection.GetPointer());
      if (josephAttenuatedForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to JosephForwardAttenuatedProjectionImageFilter."
                  << std::endl;
        return EXIT_FAILURE;
      }
      josephAttenuatedForward->SetSuperiorClipImage(superiorClipImage);
    }
    else if (args_info.fp_arg == fp_arg_MIP)
    {
      auto * mipForward = dynamic_cast<rtk::MaximumIntensityProjectionImageFilter<OutputImageType, OutputImageType> *>(
        forwardProjection.GetPointer());
      if (mipForward == nullptr)
      {
        std::cerr << "Failed to cast forward projection filter to MaximumIntensityProjectionImageFilter." << std::endl;
        return EXIT_FAILURE;
      }
      mipForward->SetSuperiorClipImage(superiorClipImage);
    }
  }
  if (args_info.sigmazero_given && args_info.fp_arg == fp_arg_Zeng)
  {
    auto * zengForward = dynamic_cast<rtk::ZengForwardProjectionImageFilter<OutputImageType, OutputImageType> *>(
      forwardProjection.GetPointer());
    if (zengForward == nullptr)
    {
      std::cerr << "Failed to cast forward projection filter to ZengForwardProjectionImageFilter." << std::endl;
      return EXIT_FAILURE;
    }
    zengForward->SetSigmaZero(args_info.sigmazero_arg);
  }
  if (args_info.alphapsf_given && args_info.fp_arg == fp_arg_Zeng)
  {
    auto * zengForward = dynamic_cast<rtk::ZengForwardProjectionImageFilter<OutputImageType, OutputImageType> *>(
      forwardProjection.GetPointer());
    if (zengForward == nullptr)
    {
      std::cerr << "Failed to cast forward projection filter to ZengForwardProjectionImageFilter." << std::endl;
      return EXIT_FAILURE;
    }
    zengForward->SetAlpha(args_info.alphapsf_arg);
  }
  forwardProjection->SetGeometry(geometry);
  if (!args_info.lowmem_flag)
  {
    TRY_AND_EXIT_ON_ITK_EXCEPTION(forwardProjection->Update())
  }

  // Write
  if (args_info.verbose_flag)
    std::cout << "Writing... " << std::endl;
  auto writer = itk::ImageFileWriter<OutputImageType>::New();
  writer->SetFileName(args_info.output_arg);
  writer->SetInput(forwardProjection->GetOutput());
  if (args_info.lowmem_flag)
  {
    writer->SetNumberOfStreamDivisions(sizeOutput[2]);
  }
  TRY_AND_EXIT_ON_ITK_EXCEPTION(writer->Update())

  return EXIT_SUCCESS;
}

File: applications/rtkforwardprojections/rtkforwardprojections.py
#!/usr/bin/env python
import sys
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = argparse.ArgumentParser(
        description="Projects a volume according to a geometry file."
    )
    parser.add_argument(
        "--verbose", "-v", help="Verbose execution", action="store_true"
    )
    parser.add_argument(
        "--geometry", "-g", help="XML geometry file name", type=str, required=True
    )
    parser.add_argument(
        "--input", "-i", help="Input volume file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output projections file name", type=str, required=True
    )
    parser.add_argument(
        "--step",
        "-s",
        help="Step size along ray (for CudaRayCast only)",
        type=float,
        default=1,
    )
    parser.add_argument(
        "--lowmem",
        "-l",
        help="Compute only one projection at a time",
        action="store_true",
    )

    # Projectors
    rtkprojectors_group = parser.add_argument_group("Projectors")
    rtkprojectors_group.add_argument(
        "--fp",
        "-f",
        help="Forward projection method",
        choices=[
            "Joseph",
            "JosephAttenuated",
            "CudaRayCast",
            "Zeng",
            "MIP",
            "CudaWrapRayCast",
        ],
        default="Joseph",
    )
    rtkprojectors_group.add_argument(
        "--attenuationmap",
        help="Attenuation map relative to the volume to perfom the attenuation correction",
        type=str,
    )
    rtkprojectors_group.add_argument(
        "--sigmazero",
        help="PSF value at a distance of 0 meter of the detector",
        type=float,
    )
    rtkprojectors_group.add_argument(
        "--alphapsf", help="Slope of the PSF against the detector distance", type=float
    )
    rtkprojectors_group.add_argument(
        "--inferiorclipimage",
        help="Value of the inferior clip of the ray for each pixel of the projections (only with Joseph-based projector)",
        type=str,
    )
    rtkprojectors_group.add_argument(
        "--superiorclipimage",
        help="Value of the superior clip of the ray for each pixel of the projections (only with Joseph-based projector)",
        type=str,
    )

    # Motion compensation options
    warp_forwardprojection_group = parser.add_argument_group(
        "Motion compensation described in [Rit et al, TMI, 2009]"
    )
    warp_forwardprojection_group.add_argument(
        "--signal", help="Signal file name", type=str
    )
    warp_forwardprojection_group.add_argument("--dvf", help="Input 4D DVF", type=str)

    rtk.add_rtk3Doutputimage_group(parser)

    return parser


def process(args_info):
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]
    OutputCudaImageType = itk.CudaImage[OutputPixelType, Dimension]

    # Geometry
    if args_info.verbose:
        print(
            f"Reading geometry information from {args_info.geometry}...",
            end="",
            flush=True,
        )
    geometry = rtk.read_geometry(args_info.geometry)
    if args_info.verbose:
        print(" done.")

    # Create a stack of empty projection images
    ConstantImageSourceType = rtk.ConstantImageSource[OutputImageType]
    constantImageSource = ConstantImageSourceType.New()
    rtk.SetConstantImageSourceFromArgParse(constantImageSource, args_info)

    # Adjust size according to geometry
    sizeOutput = list(constantImageSource.GetSize())
    sizeOutput[2] = len(geometry.GetGantryAngles())
    constantImageSource.SetSize(sizeOutput)

    # Input reader
    if args_info.verbose:
        print(f"Reading input volume {args_info.input}...")
    inputVolume = itk.imread(args_info.input)

    attenuation_map = None
    if args_info.attenuationmap:
        if args_info.verbose:
            print(f"Reading attenuation map {args_info.attenuationmap}...")
        attenuation_map = itk.imread(args_info.attenuationmap)

    inferiorClipImage = None
    superiorClipImage = None
    if args_info.inferiorclipimage:
        if args_info.verbose:
            print(f"Reading inferior clip image {args_info.inferiorclipimage}...")
        inferiorClipImage = itk.imread(args_info.inferiorclipimage)
    if args_info.superiorclipimage:
        if args_info.verbose:
            print(f"Reading superior clip image {args_info.superiorclipimage}...")
        superiorClipImage = itk.imread(args_info.superiorclipimage)

    # Create forward projection image filter
    if args_info.verbose:
        print("Projecting volume...")

    # Select the forward projector
    if args_info.fp == "Joseph":
        forwardProjection = rtk.JosephForwardProjectionImageFilter[
            OutputImageType, OutputImageType
        ].New()
    elif args_info.fp == "JosephAttenuated":
        forwardProjection = rtk.JosephForwardAttenuatedProjectionImageFilter[
            OutputImageType, OutputImageType
        ].New()
        if attenuation_map is None:
            print(
                "JosephAttenuatedForwardProjection requires an attenuation map. "
                "Please provide it using --attenuationmap."
            )
            sys.exit(1)
    elif args_info.fp == "Zeng":
        forwardProjection = rtk.ZengForwardProjectionImageFilter[
            OutputImageType, OutputImageType
        ].New()
    elif args_info.fp == "MIP":
        forwardProjection = rtk.MaximumIntensityProjectionImageFilter[
            OutputImageType, OutputImageType
        ].New()
    elif args_info.fp == "CudaRayCast":
        if hasattr(itk, "CudaImage"):
            forwardProjection = rtk.CudaForwardProjectionImageFilter[
                OutputCudaImageType
            ].New()
            forwardProjection.SetStepSize(args_info.step)
        else:
            print("The program has not been compiled with cuda option")
            sys.exit(1)
    elif args_info.fp == "CudaWrapRayCast":
        if hasattr(itk, "CudaImage"):
            if not args_info.dvf or not args_info.signal:
                print("CudaWrapRayCast requires --dvf and --signal arguments.")
                sys.exit(1)
            forwardProjection = rtk.CudaWarpForwardProjectionImageFilter[
                OutputCudaImageType
            ].New()
            forwardProjection.SetStepSize(args_info.step)

            dvf_image = itk.imread(args_info.dvf)
            forwardProjection.SetDisplacementField(dvf_image)
            forwardProjection.SetSignalFilename(args_info.signal)
        else:
            print("The program has not been compiled with cuda option")
            sys.exit(1)

    if args_info.fp in ["CudaWrapRayCast", "CudaRayCast"]:
        forwardProjection.SetInput(
            itk.cuda_image_from_image(constantImageSource.GetOutput())
        )
        forwardProjection.SetInput(1, itk.cuda_image_from_image(inputVolume))
        if attenuation_map:
            forwardProjection.SetInput(2, itk.cuda_image_from_image(attenuation_map))
    else:
        forwardProjection.SetInput(constantImageSource.GetOutput())
        forwardProjection.SetInput(1, inputVolume)
        if attenuation_map:
            forwardProjection.SetInput(2, attenuation_map)

    if inferiorClipImage is not None:
        if args_info.fp == "Joseph":
            forwardProjection.SetInferiorClipImage(inferiorClipImage)
        elif args_info.fp == "JosephAttenuated":
            forwardProjection.SetInferiorClipImage(inferiorClipImage)
        elif args_info.fp == "MIP":
            forwardProjection.SetInferiorClipImage(inferiorClipImage)
    if superiorClipImage is not None:
        if args_info.fp == "Joseph":
            forwardProjection.SetSuperiorClipImage(superiorClipImage)
        elif args_info.fp == "JosephAttenuated":
            forwardProjection.SetSuperiorClipImage(superiorClipImage)
        elif args_info.fp == "MIP":
            forwardProjection.SetSuperiorClipImage(superiorClipImage)
    if args_info.sigmazero and args_info.fp == "Zeng":
        forwardProjection.SetSigmaZero(args_info.sigmazero)
    if args_info.alphapsf and args_info.fp == "Zeng":
        forwardProjection.SetAlpha(args_info.alphapsf)

    forwardProjection.SetGeometry(geometry)

    if not args_info.lowmem:
        forwardProjection.Update()

    # Write
    if args_info.verbose:
        print("Writing...")

    writter = itk.ImageFileWriter[OutputImageType].New()
    writter.SetFileName(args_info.output)
    writter.SetInput(forwardProjection.GetOutput())
    writter.Update()

    if args_info.verbose:
        print("Processing completed successfully.")


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
